src/AutorestSwift/JsonRpc/Server.swift

```diff
@@ -0,0 +1,163 @@
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+// TODO: Update license. Reproduced from: https://github.com/apple/swift-nio-examples/tree/master/json-rpc/Sources/JsonRpc
+
+import Foundation
+import NIO
+
+public final class ChannelServer {
+    private let group: MultiThreadedEventLoopGroup
+    private let config: Config
+    private var channel: Channel?
+    private let closure: RPCClosure
+    private let logger = FileLogger(withFileName: "autorest-swift-debug.log")
+
+    public init(group: MultiThreadedEventLoopGroup, config: Config = Config(), closure: @escaping RPCClosure) {
+        self.group = group
+        self.config = config
+        self.closure = closure
+        self.state = .initializing
+    }
+
+    deinit {
+        assert(self.state == .stopped)
+    }
+
+    public func start() -> EventLoopFuture<ChannelServer> {
+        assert(state == .initializing)
+
+        let bootstrap = NIOPipeBootstrap(group: group)
+            .channelInitializer { channel in
+                channel.pipeline.addFramingHandlers(framing: .jsonpos)
+                    .flatMap {
+                        channel.pipeline.addHandler(CodableCodec<JSONRequest, JSONResponse>(), name: "AutorestServer")
+                    }.flatMap {
+                        channel.pipeline.addHandler(
+                            Handler(self.closure),
+                            name: "ServerHandler"
+                        )
+                    }
+            }.withPipes(inputDescriptor: STDIN_FILENO, outputDescriptor: STDOUT_FILENO)
+
+        state = .starting
+        return bootstrap.eventLoop.makeSucceededFuture(self)
+    }
+
+    public func stop() -> EventLoopFuture<Void> {
+        if state != .started {
+            return group.next().makeFailedFuture(ServerError.notReady)
+        }
+        guard let channel = self.channel else {
+            return group.next().makeFailedFuture(ServerError.notReady)
+        }
+        state = .stopping
+        channel.closeFuture.whenComplete { _ in
+            self.state = .stopped
+        }
+        return channel.close()
+    }
+
+    private var _state = ChannelState.initializing
+    private let lock = NSLock()
+    private var state: ChannelState {
+        get {
+            return lock.withLock {
+                _state
+            }
+        }
+        set {
+            lock.withLock {
+                _state = newValue
+                logger.log("\(self) \(_state)")
+            }
+        }
+    }
+}
+
+private class Handler: ChannelInboundHandler, RemovableChannelHandler {
+    public typealias InboundIn = JSONRequest
+    public typealias OutboundOut = JSONResponse
+
+    private let closure: RPCClosure
+
+    private let logger = FileLogger(withFileName: "autorest-swift-debug.log")
```

Comment:
Perhaps we could have a static logger on AutorestPlugin or just a singleton Logger so we don't need to have multiple logger properties?

Reply:
Changed in new iteration